### ISSUE TITLE
ForeachIteratorReconstruction drops user try/finally as disposal scaffolding while reporting Full

### ISSUE THREAD

**Issue by richlander** [richlander/dotnet-inspect#1429]
## Bug

`ForeachIteratorReconstruction` strips the iterator's compiler-synthesized enumerator-disposal scaffolding (the `<>m__FinallyN()` call, the `EndFinally`-terminated handler block, and all EH regions) by **shape alone**, with nothing that distinguishes that scaffolding from a **user-authored `try/finally`** inside the iterator. csc lowers a user `try { … } finally { … }` in an iterator into exactly the same `<>m__FinallyN()` call + `EndFinally` handler + region shape, so the user's `finally` is silently deleted during reconstruction — observable cleanup that runs on every completion (and on early iterator `Dispose()`) disappears, while the result is reported `fidelity: Full`.

## Minimal counterexample (csc-reachable, ordinary C#)

Built `-c Release`, decompiled through the cross-method import seam (the same wiring as `IteratorReconstructionPassTests.Raised`: `IrPasses.Run(fn, IrPasses.Default, context)` with `importMethodBody`):

```csharp
public static IEnumerable<int> ForeachUserFinally(IEnumerable<int> src)
{
    foreach (var x in src) { try { yield return x; } finally { Console.Write("f"); } }
}

public static IEnumerable<int> DisposeResource(IEnumerable<int> src)
{
    var d = new System.IO.StringWriter();
    try { foreach (var x in src) yield return x; } finally { d.Dispose(); }
}
```

Rendered today (both `fidelity=Full`):

```csharp
// ForeachUserFinally — the try/finally and Console.Write("f") are GONE
foreach (int item in src) { yield return item; }

// DisposeResource — d.Dispose() is GONE; the allocation is kept, so the writer silently leaks
StringWriter d = new StringWriter();
foreach (int item in src) { yield return item; }
```

A plain `foreach (var x in src) yield return x;` (no user finally) renders correctly, so the positive path is unaffected.

## Why current guards allow it

`ForeachIteratorReconstruction` (`src/ILInspector.Decompiler/Pipeline/Passes/ForeachIteratorReconstruction.cs`):

- `:114` — drops **any** block whose last child is `EndFinally` ("the fault handler … is disposal scaffolding — drop it").
- `:139-141` — drops **any** `ExpressionStatement` whose call name starts with `<>m__Finally`.
- `:157` — `work.Regions = ImmutableArray<HandlerRegion>.Empty;` clears **all** EH regions unconditionally.

None of these tie the removed region/call to the enumerator's own split-disposal (`enumeratorField.Dispose()` + the `<>7__wrap* = null` reset discovered earlier in the pass). A user `try/finally` wears the identical shape and is erased along with the disposal idiom; the foreach/yield/no-state-field validators still pass, so fidelity stays `Full`.

## Suggested fix (narrowest)

Before stripping, prove the removed region(s) and `<>m__FinallyN` are *exactly and only* the enumerator's split-disposal: require a single finally/fault whose handler body is just `enumeratorField.Dispose()` (plus the `<>7__wrap* = null` reset), tied to the same `enumeratorField` the pass already discovers, and match that one `<>m__FinallyN` to that handler instead of dropping every `<>m__Finally*` by name prefix and clearing all regions. If any additional region or any `<>m__FinallyN` whose body is not the enumerator Dispose is present, **decline** to the honest acknowledgment fallback (which lowers fidelity rather than dropping the user finally).

Adversarial fixtures: `foreach (var x in src) yield return x;` must still raise (positive canary); `foreach + try/finally` and `try { foreach … } finally { resource.Dispose(); }` must stay lowered/acknowledged (must not drop the finally).

Part of the #1356 pass over-raise correctness burndown.

**Comment by richlander** [comment]
Claiming this row (#1356 over-raise burndown). In progress — fix/issue-1429-foreach-user-finally.

**Comment by richlander** [comment]
Claimed from #1356 row 25 on branch fix/issue-1429-iterator-foreach-user-finally. Starting with the narrow disposal-scaffold discriminator and adversarial iterator try/finally fixtures.

**Comment by richlander** [comment]
Fixed on branch fix/issue-1429-foreach-user-finally → PR #1446 (In review). Gate: decline reconstruction when MoveNext references >1 distinct `<>m__Finally` helper (foreach-delegation disposal = exactly one; a user finally adds another). Verified raw: YieldEach=1 helper (still raises Full), ForeachUserFinally/DisposeResource=2 helpers (now Partial acknowledgment, finally preserved). NestedForeachYield (2 helpers, no user finally) already declined pre-change, so no positive regresses. Full suite 1176/0. GPT-5.5 adversarial review in progress.

**Comment by richlander** [comment]
Opened PR #1474. The fix gates foreach iterator reconstruction on exactly the single enumerator-disposal finally shape; extra user-finally helper calls now decline to honest iterator acknowledgment instead of being stripped.

**Comment by richlander** [comment]
Update: GPT-5.5 adversarial review (multi-round) found the initial count-based gate insufficient. The gate was rebuilt to two proof gates: (1) the enumerator must be a compiler wrapper field `<>7__wrap*` (not a hoisted user local), and (2) every `<>m__Finally` helper the body invokes must be exactly that wrapper's disposal (`enumeratorField.Dispose()` via ldfld or ldflda, storing only `<>1__state`). This now also declines a manual enumerator loop with a side-effecting Dispose and a non-IDisposable-enumerator user finally, while the YieldEach positive still raises foreach at Full. See PR #1446 for the full review writeup.
